aip/client-libraries/4222.md

```diff
@@ -154,6 +169,46 @@ If none of the routing parameters matched their respective fields, the routing h
 Much like URL parameters, if there is more than one key-value pair to be sent, the `&`
 character is used as the separator.
 
+### `path_template` syntax
+
+As seen in the above examples, the `path_template` can use a variety of symbols
+that are interpreted by code generators during conversion to regular expressions
+or non-regular expression matcher implementations. The `path_template` consists
+of segments delimited by the segment delimiter. The syntax for `path_template`
+is as follows:
+
+- The only acceptable segment delimiter is `/`.
+  - The last symbol in a `path_template` **may** be a delimiter - it will be 
+    ignored.
+- A segment **must** be of one of the following types:
+  - `*`: A single-segment wildcard. Corresponds to 1 or more non-`/` symbols.
+    The regex describing it is `[^\/]+`.
+    - A Single-segment wildcard typically represents a resource ID.
+  - `**`: A multi-segment wildcard. Corresponds to 0 or more symbols or segments
+    (so it can be empty). Consumes segment delimiters and does not match unless
```

Comment:
I think this is too confusing. There are two points to be made here:
First `**` is a multi-segment wildcard that can correspond to several things:
a. a empty string, so `foo/**` matches `foo/`
b. an empty string plus it "consumes" the delimeter immediately preceding it, so `foo/**` matches `foo`
c. one or more non-empty literal segments, so `foo/**` matches `foo/bar`, and `foo/bar/baz`, etc.

Second, the regex for the `**` changes depending on whether it's standalone or not. 
If `**` is the whole pattern, the regex is simply `.*`. 
If `**` is preceded by the other segment, then it will *include the preceding delimiter in it's regex*, and the regex for `foo/**`is `foo\/?([^\/].*)?`, note that the delimiter preceding the capture group is optional.

Reply:
Ok I tried to convey what you said but I'm getting a little confused here...

> If `**` is preceded by the other segment, then it will _include the preceding delimiter in it's regex_, and the regex for `foo/**`is `foo(\/([^\/].*)?)?`, note that the delimiter preceding the inner capture group is not optional, even though the outer capture group _as a whole_ is.

I was pulling this regex from [here](https://docs.google.com/document/d/1FtOMCBaTH7HRgBaTaWHBS8gk9phQhrPuuLoThFjGsH8/edit?resourcekey=0-_dpciY4eZCQwyh5-4sIPOA#bookmark=id.viu0ih15k0ue) (note: internal link). Is that doc not right?